tests/apps/test_train.py

```diff
@@ -0,0 +1,213 @@
+            [0.0, 1.41421356, 1.5, 0.0],
+            [1.41421356, 0.0, 0.0, 1.5],
+        ]
+    )
+    assert np.allclose(cov, target)
+
+
+@pytest.mark.usefixtures("adj", "params")
+@pytest.mark.parametrize("dim", [7, 9])  # used in the adj fixture to determine number of modes
+@pytest.mark.parametrize("n_mean", [4, 6])
+class TestVGBS:
+    """Tests for the class ``train.VGBS``"""
+
+    def test_W(self, adj, n_mean, params):
+        """Test that the W method correctly gives the diagonal matrix of embedded parameters"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        W = gbs.W(params)
+        assert np.allclose(np.diag(W), embedding(params))  # check that diagonal equals embedding
+        assert np.allclose(np.diag(np.diag(W)), W)  # check that matrix is diagonal
+
+    def test_A(self, adj, n_mean, params):
+        """Test that the A and _A_scaled methods correctly return the WAW matrix of the A_init
+        and A_init_scaled adjacency matrices, respectively, and that these outputs are different
+        (due to the scaling)"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        A = gbs.A(params)
+        A_scaled = gbs._A_scaled(params)
+        W = gbs.W(params)
+
+        assert np.allclose(A, W @ adj @ W)
+        assert np.allclose(A_scaled, W @ gbs.A_init_scaled @ W)
+        assert not np.allclose(A, A_scaled)
+
+    def test_generate_samples(self, adj, n_mean, monkeypatch):
+        """Test that _generate_samples correctly dispatches between torontonian and hafnian
+        sampling based upon whether threshold=True or threshold=False"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+
+        with monkeypatch.context() as m:
+            m.setattr(thewalrus.samples, "torontonian_sample_state", lambda *args, **kwargs: 0)
+            s_threshold = gbs._generate_samples(gbs.A_init_scaled, 10)
+
+        gbs.threshold = False
+
+        with monkeypatch.context() as m:
+            m.setattr(thewalrus.samples, "hafnian_sample_state", lambda *args, **kwargs: 1)
+            s_pnr = gbs._generate_samples(gbs.A_init_scaled, 10)
+
+        assert s_threshold == 0
+        assert s_pnr == 1
+
+    def test_add_A_init_samples_bad_shape(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples raises a ValueError when input samples of incorrect
+        shape, i.e. of dim + 1 modes"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim + 1))
+        with pytest.raises(ValueError, match="Must input samples of shape"):
+            gbs._add_A_init_samples(s)
+
+    def test_add_A_init_samples_none_there(self, adj, n_mean, dim):
+        """Test that _add_A_init_samples correctly adds samples"""
+        gbs = train.VGBS(adj, n_mean, embedding, True)
+        s = np.ones((2, dim))
+        gbs._add_A_init_samples(s)
+        assert np.allclose(gbs._A_init_samples, s)
+
+    def test_add_A_init_samples_already_there(self, adj, n_mean, dim):
```

Comment:
I'm being picky here, but technically we could be adding the wrong samples or adding them in the wrong order and this test would still pass, right? We could make initial samples to be all zeros and new samples be all ones and check that they are added as intended

Reply:
Nice that's a good point! As well as the shape, the test now checks that the first two samples are ones and the second two are zeros.